lib/tasks/multitenant/triggers.rake

```diff
@@ -38,6 +38,85 @@ namespace :multitenant do
     puts "Recreated #{triggers.size} triggers"
   end
 
+  desc 'Fix empty or corrupted tenant_id in accounts'
+  task :fix_corrupted_tenant_id_accounts, %i[batch_size sleep_time] => :environment do |_task, args|
+    batch_size = (args[:batch_size] || 100).to_i
+    sleep_time = (args[:sleep_time] || 1).to_i
+
+    ids = (Rails.application.simple_try_config_for(ENV['FILE']) || [])
+
+    ids.in_groups_of(batch_size).each do |group|
+      puts "Executing update for a batch of size: #{group.size}"
+      Account.buyers.where(id: group).update_all('tenant_id = provider_account_id') # rubocop:disable Rails/SkipsModelValidations
+      Account.providers.where(id: group).update_all('tenant_id = id') # rubocop:disable Rails/SkipsModelValidations
+      puts "Sleeping #{sleep_time} seconds"
+      sleep(sleep_time)
+    end
+  end
+
+  desc 'Fix empty or corrupted tenant_id for a table associated to account'
+  task :fix_corrupted_tenant_id_for_table_associated_to_account, %i[table_name time_start time_end batch_size sleep_time] => :environment do |_task, args|
+    sleep_time = args[:sleep_time]
+    master_id = Account.master.id
+
+    puts "------ Updating #{args[:table_name]} ------"
+    condition = condition_update_tenant_id(args[:time_start], args[:time_end])
+    args[:table_name].constantize.joining { account }.where.has(&condition).find_in_batches(batch_size: args[:batch_size].to_i) do |group|
```

Comment:
Example of DRY opportunity

```ruby
desc 'Fix empty or corrupted tenant_id for a table associated to account'
task :fix_corrupted_tenant_id_for_table_associated_to_account, %i[table_name time_start time_end batch_size sleep_time] => :environment do |_task, args|
  update_tenant_id_with_association(args) { account }
end

desc 'Fix empty or corrupted tenant_id for a table associated to user'
  task :fix_corrupted_tenant_id_for_table_associated_to_user, %i[table_name time_start time_end batch_size sleep_time] => :environment do |_task, args|sleep_time] => :environment do |_task, args|
  update_tenant_id_with_association(args) { user }
end

def update_tenant_id_with_association(args, &association_block)
  table_name = args[:table_name]
  batch_size = args[:batch_size].to_i
  sleep_time = args[:sleep_time].to_i
  condition = condition_update_tenant_id(args[:time_start], args[:time_end])
  master_id = Account.master.id

  puts "------ Updating #{table_name} ------"
  table_name.constantize.joining(&association_block).where.has(&condition).find_in_batches(batch_size: batch_size) do |group|
    puts "Executing update for a batch of size: #{group.size}"
    group.each { |record| record.update_column(:tenant_id, record.account.tenant_id) if record.account_id != master_id } # rubocop:disable Rails/SkipsModelValidations
    puts "Sleeping #{sleep_time} seconds"
    sleep(sleep_time)
  end
end
```

Reply:
the line of `record.account.tenant_id` wouldn't work to those associated to user, but I get the hint 😉 👍 thanks 💪